tests/clickhouse/optimize/test_optimize.py

```diff
@@ -244,6 +246,100 @@ def test_metrics_tags(
         assert _get_metrics_tags(table, host) == expected
 
 
+test_data = [
+    pytest.param(
+        StorageKey.ERRORS,
+        lambda dt: InsertBatch(
+            [
+                {
+                    "event_id": str(uuid.uuid4()),
+                    "project_id": 1,
+                    "group_id": 1,
+                    "deleted": 0,
+                    "timestamp": dt,
+                    "retention_days": settings.DEFAULT_RETENTION_DAYS,
+                }
+            ],
+            None,
+        ),
+        last_midnight
+        + timedelta(hours=settings.PARALLEL_OPTIMIZE_JOB_START_TIME)
+        + timedelta(minutes=30),
+        id="errors parallel",
+    )
+]
+
+
+class TestOptimizeFrequency:
+    @pytest.mark.clickhouse_db
+    @pytest.mark.redis_db
+    @pytest.mark.parametrize(
+        "storage_key, create_event_row_for_date, current_time",
+        test_data,
+    )
+    @patch("snuba.clickhouse.optimize.optimize.optimize_partitions")
+    def test_optimize_partitions_is_called_the_same_number_of_times_as_number_of_partitions(
```

Comment:
I was running this test in the debugger to better understand it, and I found that when I paused it for a while to think, the test would fail. Here's why I think that is happening:
1. By doing two inserts on a series of dates, the test creates a situation where ClickHouse has 10 partitions with 2 parts in each.
2. Left to its own devices, ClickHouse will merge the parts together, until it has 10 partitions with 1 part each.
3. If ClickHouse completes its background merge process before the test calls `optimize.get_partitions_to_optimize`, there's no optimization necessary and the test fails.

A test with a race condition like this will be flaky in CI so we need to fix it. I have two ideas on how we can do that:
1. This test asserts that if we ask `optimize_partition_runner` to schedule the work needed to optimize 10 partitions, then `optimize_partitions` is called 10 times. Can we just skip inserting all this data into the database and instead just pass `optimize_partition_runner` a list of 10 fake partitions?
2. If that doesn't work, we can tell ClickHouse to not do any background merges by adding this code to the start of the test:
```
clickhouse.execute(f"SYSTEM STOP MERGES {database}.{table}")
```

What do you think? As an aside, I suspect that this problem is why `test_optimize` is expected to flake sometimes. Can we fix it too? Probably the `SYSTEM` command is the only way to fix that one.